extensions/standard-processors/processors/TailFile.cpp

```diff
@@ -335,170 +563,210 @@ bool TailFile::storeState(const std::shared_ptr<core::ProcessContext>& context)
-      file.currentTailFilePosition_ = 0;
+  if (tail_mode_ == Mode::MULTIPLE) {
+    if (last_multifile_lookup_ + lookup_frequency_ < std::chrono::steady_clock::now()) {
+      logger_->log_debug("Lookup frequency %d ms have elapsed, doing new multifile lookup", lookup_frequency_.count());
+      checkForRemovedFiles();
+      checkForNewFiles();
+      last_multifile_lookup_ = std::chrono::steady_clock::now();
+    } else {
+      logger_->log_trace("Skipping multifile lookup");
     }
+  }
 
-    file.current_file_name_ = item.fileName;
+  // iterate over file states. may modify them
+  for (auto &state : tail_states_) {
+    processFile(context, session, state.first, state.second);
+  }
 
-    storeState(context);
+  if (!session->existsFlowFileInRelationship(Success)) {
+    yield();
   }
 }
 
-void TailFile::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
-  std::lock_guard<std::mutex> tail_lock(tail_file_mutex_);
-  std::string st_file;
-  if (context->getProperty(StateFile.getName(), st_file)) {
-    state_file_ = st_file + "." + getUUIDStr();
+void TailFile::processFile(const std::shared_ptr<core::ProcessContext> &context,
+                           const std::shared_ptr<core::ProcessSession> &session,
+                           const std::string &full_file_name,
+                           TailState &state) {
+  if (utils::file::FileUtils::file_size(full_file_name) < state.position_) {
+    processRotatedFiles(context, session, state);
   }
-  if (!this->state_recovered_) {
-    state_recovered_ = true;
-    // recover the state if we have not done so
-    this->recoverState(context);
+
+  processSingleFile(context, session, full_file_name, state);
+}
+
+void TailFile::processRotatedFiles(const std::shared_ptr<core::ProcessContext> &context,
+                                   const std::shared_ptr<core::ProcessSession> &session,
+                                   TailState &state) {
+    std::vector<TailState> rotated_file_states = findRotatedFiles(state);
+    for (TailState &file_state : rotated_file_states) {
+      processSingleFile(context, session, file_state.fileNameWithPath(), file_state);
+    }
+    state.position_ = 0;
+    state.checksum_ = 0;
+}
+
+void TailFile::processSingleFile(const std::shared_ptr<core::ProcessContext> &context,
+                                 const std::shared_ptr<core::ProcessSession> &session,
+                                 const std::string &full_file_name,
+                                 TailState &state) {
+  std::string fileName = state.file_name_;
+
+  if (utils::file::FileUtils::file_size(full_file_name) == 0u) {
+    logger_->log_warn("Unable to read file %s as it does not exist or has size zero", full_file_name);
+    return;
   }
+  logger_->log_debug("Tailing file %s from %llu", full_file_name, state.position_);
```

Comment:
Please use `PRIu64` format specifier for `uint64_t`

Reply:
done